src/lib.rs

```diff
@@ -40,19 +42,417 @@ mod repl_helper;
+        ///
+        /// # Safety
+        ///
+        /// Caller must satisfy the following preconditions for this function:
+        ///
+        /// * Valid [`Machine`] pointer created with [`machine_new`] that has not yet been freed.
+        /// * the input fulfill the safety requirements of [`CStr::From_ptr`].
+        /// * There must be no other [`QueryState`] for this [`Machine`] started by
+        /// [`run_query_iter`] that has not yet been freed with
+        /// [`query_state_free`] or the [`Machine`] state will enter an undefined
+        /// configuration with unpredictable results.
+        ///
+        /// Other concerns:
+        /// * after invoking this function, calling any other function besides [`run_query_next`]
+        /// before invoking [`query_state_free`] on the [`Machine`] pointer and [`QueryState`] pointer
+        /// will leave the [`Machine`] in an undefined state
+        ///
+        #[export_name = "scryer_run_query_iter"]
+        pub unsafe extern "C" fn run_query_iter(
+            machine: &mut Machine,
+            input: *const c_char,
+        ) -> *mut QueryState {
+            match unsafe { CStr::from_ptr(input) }.to_str() {
+                Ok(input) => {
+                    let string = input.to_string();
+                    let query_state = machine.run_query_iter(string);
+                    Box::into_raw(Box::new(query_state))
+                }
+                Err(err) => {
+                    eprintln!("Query input wasn't valid UTF-8: {err:?}");
+                    std::ptr::null_mut()
+                }
+            }
+        }
+
+        /// Cleans up the [`QueryState`] in the associated Scryer [`Machine`].
+        ///
+        /// # Safety
+        ///
+        /// * `machine` must be a valid mutable pointer to a [`Machine`] instance created by [`machine_new`] that has not yet been freed
+        /// with [`machine_free`].
+        /// * `query_state` must be a valid mutable pointer to a [`QueryState`] created by [`run_query_iter`]
+        /// that has not yet been freed by [`query_state_free`].
+        /// * The `query_state` must be from the `machine` that created it.
+        /// * There must be no other [`QueryState`] for this [`Machine`] started by
+        /// [`run_query_iter`] that has not yet been freed with
+        /// [`query_state_free`] or the [`Machine`] state will enter an undefined
+        /// configuration with unpredictable results.
+        #[export_name = "scryer_query_state_free"]
+        pub unsafe extern "C" fn query_state_free(
+            query_state: *mut QueryState,
+        ) {
+            unsafe {
+                drop(Box::from_raw(query_state));
+            }
+        }
+
+        #[export_name = "scryer_run_query_next"]
+        pub unsafe extern "C" fn run_query_next(query_state: *mut QueryState) -> *mut c_char {
+            let query_state = unsafe { &mut *query_state };
```

Comment:
Looks like I missed this earlier. I think this can take a reference instead of a pointer and then no longer needs to be unsafe.

```suggestion
        pub extern "C" fn run_query_next(query_state: &mut QueryState) -> *mut c_char {
```